crates/leon/src/parser.rs

```diff
@@ -1,249 +1,77 @@
-    },
-}
-
-impl Token {
-    fn start_text(pos: usize, ch: char) -> Self {
-        Self::Text {
-            start: pos,
-            end: pos + ch.len_utf8(),
-        }
-    }
-
-    fn start_text_single(pos: usize) -> Self {
-        Self::Text {
-            start: pos,
-            end: pos,
-        }
-    }
+impl<'s> Template<'s> {
+    pub(crate) fn parse_items(source: &'s str) -> Result<Vec<Item<'s>>, ParseError> {
+        let mut items = Vec::new();
 
-    fn start_brace_pair(pos: usize, ch: char) -> Self {
-        Self::BracePair {
-            start: pos,
-            key_seen: false,
-            end: pos + ch.len_utf8(),
-        }
-    }
+        let mut start = 0;
+        let mut s = source;
 
-    fn start_escape(pos: usize, ch: char) -> Self {
-        Self::Escape {
-            start: pos,
-            end: pos + ch.len_utf8(),
-            ch: None,
-        }
-    }
+        loop {
+            if let Some(index) = s.find(['\\', '{', '}']) {
+                if index != 0 {
+                    let (first, last) = s.split_at(index);
+                    items.push(Item::Text(first));
 
-    fn is_empty(&self, source_len: usize) -> bool {
-        match self {
-            Self::Text { start, end } => {
-                *start >= source_len || *end >= source_len || *start > *end
-            }
-            Self::BracePair {
-                start,
-                key_seen,
-                end,
-            } => !key_seen || *start >= source_len || *end >= source_len || *start > *end,
-            Self::Escape { start, end, .. } => {
-                *start >= source_len || *end >= source_len || *start > *end
-            }
-        }
-    }
+                    // Move cursor forward
+                    start += index;
+                    s = last;
+                }
+            } else {
+                if !s.is_empty() {
+                    items.push(Item::Text(s));
+                }
 
-    fn start(&self) -> usize {
-        match self {
-            Self::Text { start, .. }
-            | Self::BracePair { start, .. }
-            | Self::Escape { start, .. } => *start,
-        }
-    }
+                break Ok(items);
+            };
 
-    fn end(&self) -> usize {
-        match self {
-            Self::Text { end, .. } | Self::BracePair { end, .. } | Self::Escape { end, .. } => *end,
-        }
-    }
+            let mut chars = s.chars();
+            let ch = chars.next().unwrap();
 
-    fn set_end(&mut self, pos: usize) {
-        match self {
-            Self::Text { end, .. } | Self::BracePair { end, .. } | Self::Escape { end, .. } => {
-                *end = pos
-            }
-        }
-    }
-}
+            match ch {
+                '\\' => {
+                    match chars.next() {
+                        Some('\\' | '{' | '}') => {
+                            let t = s.get(1..2).unwrap();
+                            assert!(["\\", "{", "}"].contains(&t), "{}", t);
```

Comment:
let's either make that a `debug_assert` or remove it entirely

Reply:
Thanks for the benchmark!

I've replaced this `assert!` with `debug_assert!`.